tests/stack_test.go

```diff
@@ -524,6 +524,91 @@ func TestLocalStateLocking(t *testing.T) {
 
 }
 
+func TestLocalStateGzip(t *testing.T) { //nolint:paralleltest
+	e := ptesting.NewEnvironment(t)
+	defer func() {
+		if !t.Failed() {
+			e.DeleteEnvironment()
+		}
+	}()
+	stackName := addRandomSuffix("gzip-state")
+	integration.CreateBasicPulumiRepo(e)
+	e.ImportDirectory("integration/stack_dependencies")
+	e.SetBackend(e.LocalURL())
+	e.RunCommand("pulumi", "stack", "init", stackName)
+	e.RunCommand("yarn", "install")
+	e.RunCommand("yarn", "link", "@pulumi/pulumi")
+	e.RunCommand("pulumi", "up", "--non-interactive", "--yes", "--skip-preview")
+
+	stacksDir := filepath.Join(".pulumi", "stacks")
+	pathStack := filepath.Join(stacksDir, stackName+".json")
+	pathStackGzip := pathStack + ".gz"
+	pathStackBak := pathStack + ".bak"
+	pathStackBakGzip := pathStack + ".gz.bak"
+	sizeGzip := int64(-1)
+	sizeJSON := int64(-1)
+
+	assertStackFilesOk := func(gzip bool) {
+		gzipStackInfo, err := os.Stat(filepath.Join(e.CWD, pathStackGzip))
+		if err == nil {
+			sizeGzip = gzipStackInfo.Size()
+		}
+		jsonStackInfo, err := os.Stat(filepath.Join(e.CWD, pathStack))
+		if err == nil {
+			sizeJSON = jsonStackInfo.Size()
+		}
+
+		if gzip {
+			assert.True(t, e.PathExists(pathStackGzip), "gzip stack file ")
+			assert.True(t, e.PathExists(pathStackBakGzip), "gzip backup")
+		} else {
+			assert.True(t, e.PathExists(pathStack), "Raw json stack file")
+			assert.True(t, e.PathExists(pathStackBak), "raw backup")
+		}
+		if sizeGzip != -1 && sizeJSON != -1 {
+			assert.Greater(t, sizeJSON, sizeGzip, "Json file smaller than gzip")
+		}
+		if t.Failed() {
+			fmt.Printf("Stacks dir state at time of failure (gzip: %t):\n", gzip)
+			files, _ := ioutil.ReadDir(e.CWD + "/" + stacksDir)
+			for _, file := range files {
+				fmt.Println(file.Name(), file.Size())
+			}
+		}
+	}
+
+	// Test "pulumi up" with gzip compression on and off.
+	// Running "pulumi up" 2 times is important because normally, first the
+	// `.json` becomes `.json.gz`, then the `.json.bak` becomes `.json.gz.bak`.
+	// Default is no gzip compression
+	e.Setenv(filestate.PulumiFilestateGzipEnvVar, "0")
+	assertStackFilesOk(false)
+
+	// Enable Gzip compression
+	e.Setenv(filestate.PulumiFilestateGzipEnvVar, "1")
+	e.RunCommand("pulumi", "up", "--non-interactive", "--yes", "--skip-preview")
+	assertStackFilesOk(true)
+
+	e.RunCommand("pulumi", "up", "--non-interactive", "--yes", "--skip-preview")
+	assertStackFilesOk(true)
+
+	// Disable Gzip compression
+	e.Setenv(filestate.PulumiFilestateGzipEnvVar, "0")
+	e.RunCommand("pulumi", "up", "--non-interactive", "--yes", "--skip-preview")
+	assertStackFilesOk(false)
```

Comment:
Am I reading this right that once a user opts into gzip, disabling causes a failure?

Reply:
No, this checks that when we re-run `pulumi up` with gzip compression disabled, pulumi reads the stack (whether `.json.gz` or `.json`) and writes `.json` (check the definition of `assertStackFilesOk`).

Reply:
@iwahbe: A nit then on renaming `assertStackFilesOk` to something like `assertStackFileFormat` and taking a string argument.